node/core/provisioner/src/lib.rs

```diff
@@ -0,0 +1,222 @@
+			job.run_loop().await
+		}
+		.boxed()
+	}
+}
+
+impl ProvisioningJob {
+	pub fn new(relay_parent: Hash, sender: mpsc::Sender<FromJob>, receiver: mpsc::Receiver<ToJob>) -> Self {
+		Self {
+			relay_parent,
+			sender,
+			receiver,
+			provisionable_data_channels: Vec::new(),
+			backed_candidates: Vec::new(),
+			signed_bitfields: Vec::new(),
+		}
+	}
+
+	async fn run_loop(mut self) -> Result<(), Error> {
+		while let Some(msg) = self.receiver.next().await {
+			use ProvisionerMessage::{RequestBlockAuthorshipData, RequestInherentData, ProvisionableData};
+
+			match msg {
+				ToJob::Provisioner(RequestInherentData(_, sender)) => self.send_inherent_data(sender).await?,
+				ToJob::Provisioner(RequestBlockAuthorshipData(_, sender)) => self.provisionable_data_channels.push(sender),
+				ToJob::Provisioner(ProvisionableData(data)) => {
+					for channel in self.provisionable_data_channels.iter_mut() {
+						// REVIEW: the try operator here breaks the run loop if any receiver ever unexpectedly
+						// closes their channel. Is that desired?
+						channel.send(data.clone()).await?;
+					}
+					self.note_provisionable_data(data);
+				}
+				ToJob::Stop => break,
+			}
+		}
+
+		Ok(())
+	}
+
+	// REVIEW: not async, but small; hopefully permissable
+	fn note_provisionable_data(&mut self, provisionable_data: ProvisionableData) {
+		use ProvisionableData::{Bitfield, BackedCandidate};
+
+		match provisionable_data {
+			Bitfield(_, signed_bitfield) => self.signed_bitfields.push(signed_bitfield),
+			BackedCandidate(backed_candidate) => self.backed_candidates.push(backed_candidate),
+			_ => {}
+		}
+	}
+
+	// The provisioner is the subsystem best suited to choosing which specific
+	// backed candidates and availability bitfields should be assembled into the
+	// block. To engage this functionality, a
+	// `ProvisionerMessage::RequestInherentData` is sent; the response is a set of
+	// non-conflicting candidates and the appropriate bitfields. Non-conflicting
+	// means that there are never two distinct parachain candidates included for
+	// the same parachain and that new parachain candidates cannot be included
+	// until the previous one either gets declared available or expired.
+	//
+	// The main complication here is going to be around handling
+	// occupied-core-assumptions. We might have candidates that are only
+	// includable when some bitfields are included. And we might have candidates
+	// that are not includable when certain bitfields are included.
+	//
+	// When we're choosing bitfields to include, the rule should be simple:
+	// maximize availability. So basically, include all bitfields. And then
+	// choose a coherent set of candidates along with that.
+	async fn send_inherent_data(&mut self, sender: oneshot::Sender<ProvisionerInherentData>) -> Result<(), Error> {
+
+		// coherency rules for backed candidates, recap:
+		//
+		// - [x] only one per parachain
+		// - [ ] candidate is available (2/3+ of validators have set the 1 bit)
```

Comment:
We don't care at this stage if the candidate is available. The availability process only kicks off once the candidate is backed (_in the chain_). What we need to make sure of is that we only include candidates that are possible to include.

I recommend revisiting the documentation on the `Scheduler`, particularly the section on availability cores.

And note the information that the `availability_cores` runtime API provides. 

So basically, we need to see which core this candidate can be included on. If its para is on a `ScheduledCore`, then it is unconditionally includable. Otherwise, we have two cases to deal with:
  * If it is `next_up_on_available` on an `OccupiedCore`, we can only include the candidate if the bitfields we are including _and_ the `availability` vec of the `OccupiedCore` together would form 2/3+ of validators.
  * If it is `next_up_on_time_out` on an `OccupiedCore`, we can only include the candidate if the bitfields we are including _and_ the `availability` vec of the `OccupiedCore` together would form <2/3 of validators _and_ the `time_out_at` is the block we are building.

Otherwise, the candidate's para isn't represented anywhere in the set of cores and it is thus not scheduled and we can't include.

------

I note that the above logic is very complicated and the simplest thing for the initial version is to have a basic rule: include the candidate if there is a `Scheduled` core for its para, and don't mess around with bitfields yet. We will do these optimizations for optimistic inclusion later.

Reply:
And also note that candidates are to be ordered ascending by core index.